datafusion/src/datasource/listing/table.rs

```diff
@@ -118,12 +195,17 @@ impl ListingTable {
     /// The provided `schema` must be resolved before creating the table
     /// and should contain the fields of the file without the table
     /// partitioning columns.
-    pub fn new(
-        object_store: Arc<dyn ObjectStore>,
-        table_path: String,
-        file_schema: SchemaRef,
-        options: ListingOptions,
-    ) -> Self {
+    pub async fn new(mut config: ListingTableConfig) -> Result<Self> {
```

Comment:
I like this API -- it would be cool if we could add some more documentation (either here or in `ListingTableConfig` explaining that the options / schema are inferred if not explicitly created

Also, as this can now return an `Err` perhaps renaming to `pub async fn try_new(...)` would be a more idiomatic / clearer name